### PR TITLE
Unregister page's service worker before auditing it.

### PR DESCRIPTION
**Pull request by paulirish**
Currently we just flip on "force upload on load" which isn't a mode that's available outside of devtools. 

This will remove the worker completely, and allow it to register and function as it does in the wild.

### CODE REVIEW

File: lighthouse-core/gather/drivers/driver.js [hunk 1]
```diff
@@ -386,14 +391,51 @@ class Driver {
...
+
+    return Promise.all([getVersions, getRegistrations]).then(res => {
+      const versions = res[0].versions;
+      const registrations = res[1].registrations;
+      let workerVersionsToStop = [];
+      let registrationsToUnregister = [];
+
+      registrations
+        // find any non-deleted registrations where the domain matches the one entered
+        // CAVEAT: this will not match if a redirect changes domain.com to m.domain.com
```
Comment (paulirish) [inline lighthouse-core/gather/drivers/driver.js]: I'll be able to nuke this caveat once #582 is in.

File: lighthouse-core/gather/drivers/driver.js [hunk 2]
```diff
@@ -105,6 +106,10 @@ class Driver {
     // log event listeners being bound
     this.formattedLog('listen for event =>', {method: eventName}, 'verbose');
     this._eventEmitter.on(eventName, cb);
+    this._eventEmitter.on(eventName, data => {
```
Comment (brendankenny) [inline lighthouse-core/gather/drivers/driver.js]: meant to delete the other `cb` registration?

File: lighthouse-core/gather/drivers/driver.js [hunk 3]
```diff
@@ -386,14 +391,51 @@ class Driver {
...
 
-  forceUpdateServiceWorkers() {
-    // COMPAT: This command will trigger this registrationId error in Chrome 50 (51 undetermined):
-    //   "{"code":-32602,"message":"Missing or invalid 'registrationId' parameter"}"
-    return this.sendCommand('ServiceWorker.setForceUpdateOnPageLoad', {
-      forceUpdateOnPageLoad: true
+  unregisterServiceWorker(pageUrl) {
+    // first, get data about active SW versions and registrations
+    const getVersions = new Promise(resolve => {
+      this.once('ServiceWorker.workerVersionUpdated', data => resolve(data));
```
Comment (brendankenny) [inline lighthouse-core/gather/drivers/driver.js]: can do `this.once('ServiceWorker.workerVersionUpdated', resolve);`

File: lighthouse-core/gather/drivers/driver.js [hunk 4]
```diff
@@ -386,14 +391,51 @@ class Driver {
...
+    });
+    const getRegistrations = new Promise(resolve => {
+      this.once('ServiceWorker.workerRegistrationUpdated', data => resolve(data));
```
Comment (brendankenny) [inline lighthouse-core/gather/drivers/driver.js]: same here

**Comment by paulirish** [comment]
### [`Storage.clearDataForOrigin`](https://chromedevtools.github.io/debugger-protocol-viewer/tot/Storage/#method-clearDataForOrigin) and heavy duty Service Worker removal

TIL this protocol method exists. :D  I poked into it a bit. It hooks into the "clear site data" functionality that is exposed in Chrome preferences.  It's pretty sweet.

As for service worker, it appears to unregister the registration.  This doesn't stop any active worker (which my PR does). However, I'm not sure this distinction matters.
On https://caltrainschedule.io/ I can trigger clearDataForOrigin which swaps the worker into an "Unregistered but Running" state. At this point, I can disable my wifi and open the url in a new tab. The unregistered SW will come back from the dead to serve the page. This is also true if I _Stop_ the worker before Unregistering things.

So what this means is that both my PR and the clearDataForOrigin both fail to remove an active SW entirely. The requirement is really that all tabs aka clients of the service worker must be closed for the unregistration to be successful. 

When all tabs are closed, and the machine offline, if I reopen the tab, the page will actually still load, and a new sw will enter into the "starting" phase.  However a hard reload will eventually get me a dinosaur. (I think a clear cache would prevent this last bit from happening.)

So tldr: SW's be cray. Gotta close all your tabs if you want an unregistration to be successful. And we should probably adopt clearDataForOrigin (rather than this PR) while we figure out a way in LH to deal with extra tabs open that prevent clean testing of a SW.

**Comment by wardpeet** [comment]
@paulirish are you going to look into clearDataForOrigin? or just let it rest a bit until more info is available?

**Comment by brendankenny** [comment]
And should we close this PR or let it lie in state for a little longer? :)

For the open tabs, in the extension at least, we can do a [tab query for the domain](https://developer.chrome.com/extensions/tabs#property-queryInfo-url) and warn specifically that other tabs are open that won't allow the SW to be removed.

I don't see anything that would do the equivalent for the CLI, but about:inspect gets a list of inspectable tabs and their URLs, so it seems like it should be possible...

**Comment by paulirish** [comment]
>  don't see anything that would do the equivalent for the CLI,

CRI has a [`List` method](https://github.com/cyrus-and/chrome-remote-interface#modulelistoptions-callback) to get all tabs. So we're good.
image


-------------

pouring one out for my PR and closing..
